### PR TITLE
Mqtt client implementation

### PR DESCRIPTION
**Pull request by bmaidics**

### CODE REVIEW

File: incubator/binding-mqtt.spec/src/main/scripts/io/aklivity/zilla/specs/binding/mqtt/streams/application/subscribe.receive.message.overlapping.wildcard/server.rpt [hunk 1]
```diff
@@ -18,6 +18,67 @@ accept "zilla://streams/app0"
...
+                                .kind("STATE")
+                                .build()
+                             .build()}
+
+read ${mqtt:session()
+               .subscription("sensor/+/#", 1)
+               .subscription("sensor/+/1", 2)
+               .build()}
+
+write ${mqtt:session()
```
Comment (jfallows) [inline incubator/binding-mqtt.spec/src/main/scripts/io/aklivity/zilla/specs/binding/mqtt/streams/application/subscribe.receive.message.overlapping.wildcard/server.rpt:75]: `write zilla:data.ext ...`?

File: incubator/binding-mqtt.spec/src/main/scripts/io/aklivity/zilla/specs/binding/mqtt/streams/application/connect.reject.will.retain.not.supported/client.rpt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                                .packetSizeMax(66560)
+                                .capabilities("WILDCARD", "SUBSCRIPTION_IDS", "SHARED_SUBSCRIPTIONS")
+                                .clientId("client")
+                                .build()
+                              .build()}
+
+connected
+
+read zilla:data.empty
+
```
[the remark this answers is not in the document]
Comment (jfallows) [reply incubator/binding-mqtt.spec/src/main/scripts/io/aklivity/zilla/specs/binding/mqtt/streams/application/connect.reject.will.retain.not.supported/client.rpt:43]: same feedback for other "reject" scenarios where session stream needs to clean up.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/MqttConfiguration.java [hunk 3]
```diff
@@ -32,7 +33,6 @@ public class MqttConfiguration extends Configuration
     public static final BooleanPropertyDef WILDCARD_SUBSCRIPTION_AVAILABLE;
     public static final BooleanPropertyDef SUBSCRIPTION_IDENTIFIERS_AVAILABLE;
     public static final BooleanPropertyDef SHARED_SUBSCRIPTION_AVAILABLE;
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/MqttConfiguration.java:35]: Can we now remove these `..._AVAILABLE` properties in `MqttConfiguration`?

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java [hunk 4]
```diff
@@ -1251,6 +1237,9 @@ private final class MqttServer
         private long keepAliveTimeoutId = NO_CANCEL_ID;
         private long keepAliveTimeoutAt;
 
+        private int maximumQos;
+        private int packetSizeMax;
+        private int capabilities = 7;
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:1242]: Use the mask constants to define default capabilities.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java [hunk 5]
```diff
@@ -1880,36 +1879,55 @@ private void onDecodePublish(
...
+                        publishBuilder.qos(mqttPublishHeaderRO.qos);
+                        publishBuilder.flags(mqttPublishHeaderRO.flags);
+                        publishBuilder.expiryInterval(mqttPublishHeaderRO.expiryInterval);
+                        publishBuilder.contentType(mqttPublishHeaderRO.contentType);
+                        publishBuilder.format(f -> f.set(mqttPublishHeaderRO.payloadFormat));
+                        publishBuilder.responseTopic(mqttPublishHeaderRO.responseTopic);
+                        publishBuilder.correlation(c -> c.bytes(mqttPublishHeaderRO.correlationData));
+                        final Array32FW<MqttUserPropertyFW> userProperties = userPropertiesRW.build();
+                        userProperties.forEach(c -> publishBuilder.propertiesItem(p -> p.key(c.key()).value(c.value())));
+                    });
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:1915]: Rename `p` to `pi` (properties item), `publishBuilder` to `p` (lambda variable) and use fluent style of method chaining to set `qos`, `flags`, ... in a single statement as we usually do.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java [hunk 6]
```diff
@@ -1880,36 +1879,55 @@ private void onDecodePublish(
...
+                doSignalKeepAliveTimeout();
             }
-            doSignalKeepAliveTimeout();
         }
 
-        private void onDecodeSubscribe(
+        private int onDecodeSubscribe(
             long traceId,
             long authorization,
+            int progress,
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:1930]: What is triggering this change?

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java [hunk 7]
```diff
@@ -1945,78 +1964,86 @@ private void onDecodeSubscribe(
...
             {
                 final List<Subscription> newSubscriptions = new ArrayList<>();
 
-                for (int decodeProgress = decodeOffset; decodeProgress < decodeLimit; )
+                decode:
                 {
-                    final MqttSubscribePayloadFW mqttSubscribePayload =
-                        mqttSubscribePayloadRO.tryWrap(decodeBuffer, decodeProgress, decodeLimit);
-                    if (mqttSubscribePayload == null)
+                    //We don't know the server capabilities yet
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:1975]: Move this comment inside the `if` above the `break decode;`

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java [hunk 8]
```diff
@@ -2064,6 +2093,17 @@ private void openSubscribeStreams(
         {
             final Long2ObjectHashMap<List<Subscription>> subscriptionsByRouteId = new Long2ObjectHashMap<>();
 
+            if (!adminSubscribe)
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:2096]: The idea of `adminSubscribe` is good, but `admin` is only one use case and has no subscription id, and we've discovered we need to handle this for non-admin subscriptions during reconnect.

Suggest renaming to `implicitSubscribe` or similar to cover both use cases; admin and reconnect.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java [hunk 9]
```diff
@@ -2083,6 +2123,7 @@ private void openSubscribeStreams(
                 MqttSubscribeStream stream = subscribeStreams.computeIfAbsent(subscribeKey, s ->
                     new MqttSubscribeStream(routedId, key, adminSubscribe));
                 stream.packetId = packetId;
+                value.removeIf(s -> s.reasonCode > GRANTED_QOS_2);
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:2126]: Why is this needed?

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java [hunk 10]
```diff
@@ -2171,9 +2214,10 @@ private void doSendSessionState(
             newState.forEach(subscription ->
                 sessionStateBuilder.subscriptionsItem(subscriptionBuilder ->
                 {
-                    subscriptionBuilder.pattern(subscription.filter);
                     subscriptionBuilder.subscriptionId(subscription.id);
+                    subscriptionBuilder.qos(subscription.qos);
                     subscriptionBuilder.flags(subscription.flags);
+                    subscriptionBuilder.pattern(subscription.filter);
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:2220]: Rename `subscriptionBuilder` to `s` (lambda variable) and use fluent style to set `pattern`, `subscriptionId`, etc in a single statement.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java [hunk 11]
```diff
@@ -3619,7 +3658,7 @@ private void doPublishBegin(
                         {
                             publishBuilder.clientId(clientId);
                             publishBuilder.topic(topic);
-                            publishBuilder.flags(retainedMessages);
+                            publishBuilder.flags(retainAvailable(capabilities) ? 1 : 0);
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:3661]: Rename `publishBuilder` to `p` and use fluent style.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java [hunk 12]
```diff
@@ -15,6 +15,4122 @@
...
+    private final SignalFW signalRO = new SignalFW();
+
+    private final BeginFW.Builder beginRW = new BeginFW.Builder();
+    private final DataFW.Builder dataRW = new DataFW.Builder();
+    private final EndFW.Builder endRW = new EndFW.Builder();
+    private final AbortFW.Builder abortRW = new AbortFW.Builder();
+    private final WindowFW.Builder windowRW = new WindowFW.Builder();
+    private final ResetFW.Builder resetRW = new ResetFW.Builder();
+    private final FlushFW.Builder flushRW = new FlushFW.Builder();
+
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:236]: Remove extra blank line.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java [hunk 13]
```diff
@@ -15,6 +15,4122 @@
...
+    private final LongFunction<BudgetDebitor> supplyDebitor;
+    private final Long2ObjectHashMap<MqttBindingConfig> bindings;
+    private final int mqttTypeId;
+
+    private final long publishTimeoutMillis;
+    private final long connackTimeoutMillis;
+    private final int encodeBudgetMax;
+
+    private final CharsetDecoder utf8Decoder;
+
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:346]: Remove extra blank line.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java [hunk 14]
```diff
@@ -15,6 +15,4122 @@
...
+        {
+            final long resolvedId = resolved.id;
+
+            final OctetsFW extension = begin.extension();
+            final ExtensionFW beginEx = extension.get(extensionRO::tryWrap);
+            assert beginEx != null;
+            final int typeId = beginEx.typeId();
+            assert typeId == mqttTypeId;
+
+
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:427]: Remove extra blank line.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java [hunk 15]
```diff
@@ -15,6 +15,4122 @@
...
+    }
+
+    private MqttClient resolveClient(
+        long routedId,
+        long resolvedId,
+        long initialId,
+        String16FW clientId)
+    {
+        final int clientKey = clientKey(clientId.asString());
+        return clients.computeIfAbsent(clientKey,
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:466]: Avoid using `.intern()` on strings that can be provided by external untrusted clients.
Suggest using `Object2ObjectHashMap` for now, and later switch to using `affinity` (see core.idl) instead of passing `clientId` in `publish` and `subscribe` streams extensions.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java [hunk 16]
```diff
@@ -15,6 +15,4122 @@
...
+            {
+                client.onDecodeError(traceId, authorization, reasonCode);
+                client.decoder = decodeIgnoreAll;
+            }
+        }
+
+        return progress;
+    }
+
+    private boolean invalidUtf8(OctetsFW payload)
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:1031]: Paramaeter formatting.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java [hunk 17]
```diff
@@ -15,6 +15,4122 @@
...
+    {
+        boolean invalid = false;
+        byte[] payloadBytes = new byte[payload.sizeof()];
+        payload.value().getBytes(0, payloadBytes);
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:1035]: Avoid allocation when verifying UTF-8.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java [hunk 18]
```diff
@@ -15,6 +15,4122 @@
...
+            long authorization,
+            long budgetId,
+            DirectBuffer buffer,
+            int offset,
+            int limit);
+    }
+
+    private final class MqttClient
+    {
+        //TODO: organize final and non-final fields
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:1163]: Complete TODO

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java [hunk 19]
```diff
@@ -15,6 +15,4122 @@
...
+            String topic)
+        {
+            return sessionStream.subscriptions.stream().anyMatch(s ->
+            {
+                String regex = s.filter.replace("#", ".*").replace("+", "[^/]+");
+                return Pattern.matches(regex, topic);
+            });
+        }
+
+        private int getNextPacketId()
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:2700]: Rename to `nextPacketId()`.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java [hunk 20]
```diff
@@ -15,6 +15,4122 @@
...
+            {
+                reasonCode = mqttPublishHeader.decode(client, publish.topicName(), publish.properties(), publish.typeAndFlags());
+            }
+
+            if (reasonCode == SUCCESS)
+            {
+                final int qos = mqttPublishHeader.qos;
+                MqttSubscribeStream subscriber = client.subscribeStreams.get(qos);
+
+                if (!client.existStreamForTopic(mqttPublishHeader.topic))
```
Comment (jfallows) [inline incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:958]: This check feels heavy for every message received at the mqtt client.
The topic has been decoded via UTF-8 into String and the matcher is reconstructed on every call.
Can we use subscription id to help make this more efficient for the typical case?

**Review by jfallows — CHANGES_REQUESTED** [review]

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java (hunk 6 above)
Comment (bmaidics) [reply incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:1930]: We don't want to progress decoding subscribes until we have the capabilities, as we don't know if we need to refuse wildcard/shared subscriptions etc.

Comment (bmaidics) [reply incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:1930]: So we want to buffer the subscribe in the decodebuffer until we receive the capabilitie

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java (hunk 9 above)
Comment (bmaidics) [reply incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java:2126]: We don't want to send back to the subscribe stream that "these are the new subscription states" if we received back non-successful suback for that subscription. But we need to receive back these from the sessionStream in order to delegate the SUBACK. So we're removing the non-successful ones when sending the new subscribe filters.

File: incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java (hunk 15 above)
Comment (bmaidics) [reply incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttClientFactory.java:466]: Use affinity, and Int2ObjectHashmap

**Review by jfallows — APPROVED** [review]
